File: src/blackjack.js
(async() => {
    const program = "src/blackjack.pl";
    
    // Creamos la sesión de Prolog y cargamos la base de conocimientos
    const session = pl.create();
    await session.promiseConsult(program);

    // Obtenemos las cartas iniciales
    let [playerCards, dealerCards, restDeck] = await getInitialCards(session);
    console.log(playerCards, dealerCards);

    // Get canvas
    canvas = document.getElementById("game")
    ctx = canvas.getContext("2d")

    // Load card spritesheet
    cardImage = new Image()
    cardImage.src = "images/playingCards.png"
    cardImage.onload = function() {
        drawCards(cardsFromList(playerCards), cardsFromList(dealerCards), true)
    }

    // Get buttons
    const hitButton = document.getElementById("hit")
    const standButton = document.getElementById("stand")

    // Add event listeners
    hitButton.addEventListener("click", async function() {
        [playerCards, restDeck] = await playerDrawCard(session, restDeck, playerCards);
        ctx.clearRect(0, 0, canvas.width, canvas.height);
        drawCards(cardsFromList(playerCards), cardsFromList(dealerCards), true)

        // Check if player has busted
        const player_busted = await playerHasBusted(session, playerCards);
        if(player_busted) {
            ctx.fillText("¡Te has estallado!", 15, 400)
            hitButton.disabled = true;
            standButton.disabled = true;
        }
    })

    standButton.addEventListener("click", async function() {
        dealerCards = await dealerTurn(session, restDeck, dealerCards);
        ctx.clearRect(0, 0, canvas.width, canvas.height);
        drawCards(cardsFromList(playerCards), cardsFromList(dealerCards), false)

        const winner = await determineWinner(session, playerCards, dealerCards);
        if(winner.id === "player")
            ctx.fillText("¡Has ganado!", 15, 400)
        else if(winner.id === "dealer")
            ctx.fillText("¡Has perdido!", 15, 400)
        else
            ctx.fillText("¡Has empatado!", 15, 400)
        
        hitButton.disabled = true;
        standButton.disabled = true;
    })
})();

// Determina si el jugador se ha pasado o no
async function playerHasBusted(session, hand) {
    const goal = "has_player_busted(" + hand.toString() + ").";
    await session.promiseQuery(goal);
    const answer = await session.promiseAnswer();
    return answer;
}

// El jugador pide una carta
async function playerDrawCard(session, deck, playerHand) {
    const goal = "draw_card_into_hand(" + deck.toString() + ", NewDeck," + playerHand.toString() + ", NewHand).";
    await session.promiseQuery(goal);
    const answer = await session.promiseAnswer();
    const newHand = answer.lookup("NewHand");
    const rest = answer.lookup("NewDeck");
    return [newHand, rest];
}

// Hace que el crupier juegue su turno
async function dealerTurn(session, deck, dealerHand) {
    const goal = "dealer_turn(" + deck.toString() + ", " + dealerHand.toString() + ", NewDealerHand).";
    await session.promiseQuery(goal);
    const answer = await session.promiseAnswer();
    const newHand = answer.lookup("NewDealerHand");
    return newHand;
}

// Determina el ganador de la partida
async function determineWinner(session, playerHand, dealerHand) {
    const goal = "determine_winner(" + playerHand.toString() + ", " + dealerHand.toString() + ", Winner).";
    await session.promiseQuery(goal);
    const answer = await session.promiseAnswer();
    const winner = answer.lookup("Winner");
    return winner;
}

// Obtiene las cartas iniciales del jugador y del crupier, así como el mazo restante
async function getInitialCards(session) {
    const goal = "initial_setup(PlayerCards, DealerCards, RestDeck).";
    await session.promiseQuery(goal);
    const answer = await session.promiseAnswer();
    const playerCards = answer.lookup("PlayerCards");
    const dealerCards = answer.lookup("DealerCards");
    const restDeck = answer.lookup("RestDeck");
    return [playerCards, dealerCards, restDeck];
}

// Convierte una carta de Prolog a un objeto de carta
function convertCard(card) {
    return {
        suit: card.args[0].toString(),
        value: card.args[1].toString()
    }
}

// Convierte una lista de cartas de Prolog a un array de objetos de carta
function cardsFromList(xs) {
    var arr = [];
    while(pl.type.is_term(xs) && xs.indicator === "./2") {
        arr.push(convertCard(xs.args[0]));
        xs = xs.args[1];
    }
    if(pl.type.is_term(xs) && xs.indicator === "[]/0")
        return arr;
    return null;
}

// Dibuja una carta en el canvas
function drawCard(card, x, y) {
    // Determinamos el orden de las cartas en el spritesheet
    const suitOrder = ["hearts", "diamonds", "spades", "clubs", "back"]
    const cardOrder = ["two", "three", "four", "five", "six", "seven", "eight", "nine", "ten", "jack", "queen", "king", "ace"]
    // Según la carta que sea, obtenemos su posición en el spritesheet
    const suitIndex = suitOrder.indexOf(card.suit)
    const cardIndex = cardOrder.indexOf(card.value)
    // Coordenadas de la carta en el spritesheet
    var cardWidth = 81, cardHeight = 117
    var cardX = cardWidth * cardIndex
    var cardY = cardHeight * suitIndex
    // Dibujamos la carta en el canvas
    ctx.drawImage(cardImage, cardX, cardY, cardWidth, cardHeight, x, y, cardWidth, cardHeight)
}

// Dibuja las cartas del jugador y del crupier en el canvas
function drawCards(playerCards, dealerCards, hideFirstCard) {
    // Establecemos la fuente
    ctx.font = "30px Arial"

    // Dibujamos las cartas del jugador
    ctx.fillText("Tú", 15, 40)
    for(var i = 0; i < playerCards.length; i++) {
        drawCard(playerCards[i], 15 + i * 100, 55)
    }

    // Dibujamos las cartas del crupier
    ctx.fillText("Crupier", 15, 220)
    for(var i = 0; i < dealerCards.length; i++) {
        if(i == 0 && hideFirstCard) {
            drawCard({suit: "back", value: "two"}, 15 + i * 100, 235)
        } else {
            drawCard(dealerCards[i], 15 + i * 100, 235)
        }
    }
}

File: src/blackjack.pl
:- use_module(library(random)).
:- use_module(library(lists)).

% cards/2 devuelve todas las cartas en una baraja de cartas
cards(Deck) :-
    Suits = [spades, hearts, clubs, diamonds],
    Cards = [ace, two, three, four, five, six, seven, eight, nine, ten, jack, queen, king],
    findall(card(Suit, Card), (member(Suit, Suits), member(Card, Cards)), Deck).

% value/2 devuelve el valor de una carta
value(card(_, ace), 11).
value(card(_, two), 2).
value(card(_, three), 3).
value(card(_, four), 4).
value(card(_, five), 5).
value(card(_, six), 6).
value(card(_, seven), 7).
value(card(_, eight), 8).
value(card(_, nine), 9).
value(card(_, ten), 10).
value(card(_, jack), 10).
value(card(_, queen), 10).
value(card(_, king), 10).

% shuffle/2 baraja una lista de cartas de manera aleatoria
shuffle(Cards, Shuffled) :-
    random_permutation(Cards, Shuffled).

% draw_card/2 elimina la primera carta de una baraja y la devuelve
draw_card([Card | Rest], Card, Rest).

% draw_card_into_hand/3 agrega una carta a una mano de cartas
draw_card_into_hand(Deck, NewDeck, Hand, NewHand) :-
    draw_card(Deck, Card, NewDeck),
    append(Hand, [Card], NewHand).

% score/2 calcula el puntaje de una mano de cartas
score(Hand, Score) :-
    score(Hand, 0, Score).

% score/3 es un predicado auxiliar para score/2
score([], Acc, Acc).
score([card(_, Type) | Rest], Acc, Score) :-
    value(card(_, Type), Value),
    (Type = ace, Acc + 11 > 21 ->
        NewAcc is Acc + 1
    ;   NewAcc is Acc + Value),
    score(Rest, NewAcc, Score).

% initial_setup/3 inicializa el juego de blackjack repartiendo dos cartas a cada jugador
% y devuelve las cartas de cada jugador y el resto de la baraja
initial_setup(PlayerCards, DealerCards, Rest4) :-
    cards(Deck),
    shuffle(Deck, ShuffledDeck),
    draw_card(ShuffledDeck, PlayerCard1, Rest1),
    draw_card(Rest1, DealerCard1, Rest2),
    draw_card(Rest2, PlayerCard2, Rest3),
    draw_card(Rest3, DealerCard2, Rest4),
    PlayerCards = [PlayerCard1, PlayerCard2],
    DealerCards = [DealerCard1, DealerCard2].

% dealer_turn/2 juega el turno del crupier hasta que su puntaje sea mayor o igual a 17
dealer_turn(Deck, DealerCards, NewDealerCards) :-
    score(DealerCards, DealerSum),
    DealerSum < 17,
    draw_card_into_hand(Deck, NewDeck, DealerCards, NewDealerCards),
    dealer_turn(NewDeck, NewDealerCards, NewDealerCards).
dealer_turn(Deck, DealerCards, DealerCards).

% determine_winner/3 determina el ganador del juego
determine_winner(PlayerCards, DealerCards, Winner) :-
    score(PlayerCards, PlayerScore),
    score(DealerCards, DealerScore),
    (PlayerScore > 21 ->
        Winner = dealer
    ;   DealerScore > 21 ->
        Winner = player
    ;   PlayerScore > DealerScore ->
        Winner = player
    ;   DealerScore > PlayerScore ->
        Winner = dealer
    ;   Winner = push).

% has_player_busted/1 devuelve true si el jugador se pasó de 21
has_player_busted(PlayerCards) :-
    score(PlayerCards, PlayerSum),
    PlayerSum > 21.